### PR TITLE
0_RootFS: Support riscv64

### PR DESCRIPTION
**Pull request by eschnett**

### CODE REVIEW

File: 0_RootFS/gcc_sources.jl [hunk 1]
```diff
@@ -236,7 +236,7 @@ function gcc_sources(gcc_version::VersionNumber, compiler_target::Platform; kwar
                 ArchiveSource("https://mirrors.kernel.org/gnu/glibc/glibc-2.19.tar.xz",
                               "2d3997f588401ea095a0b27227b1d50cdfdd416236f6567b564549d3b46ea2a2"),
             ]
-        elseif arch(compiler_target) in ["x86_64", "i686", "powerpc64le"]
+        elseif arch(compiler_target) in ["x86_64", "i686", "powerpc64le", "riscv64"]
```
Comment (giordano) [inline 0_RootFS/gcc_sources.jl:239]: As mentioned in https://github.com/JuliaPackaging/BinaryBuilderBase.jl/pull/375#discussion_r1867458878, this doesn't make sense.  We should use a different glibc version for risc-v, maybe something like 2.36 or so, to target systems we have access to.

Comment (eschnett) [reply 0_RootFS/gcc_sources.jl:239]: Thanks. I was trying 2.28; switching to 2.36 now.

Comment (giordano) [reply 0_RootFS/gcc_sources.jl:239]: Hold on a few minutes, I'm trying other RISC-V I have access to, I seem to have different glibc versions there.

Comment (giordano) [reply 0_RootFS/gcc_sources.jl:239]: Ok, some of the boards in https://riscv.epcc.ed.ac.uk/status/ use Ubuntu 22.04 and hence have Glibc 2.35.  Also, according to [Repology](https://repology.org/project/glibc/versions) Red Hat 9 and derivatives (AlmaLinx/Rocky Linux) ship with Glibc 2.34, but my understanding is that Red Hat doesn't currently support RISC-V, so we probably shouldn't care about that at the moment.

To summarise, given the fact we have boards with Glibc 2.35 and Ubuntu 22.04 looks a reasonable baseline for a somewhat recent architecture I think we should target 2.35

Comment (eschnett) [reply 0_RootFS/gcc_sources.jl:239]: Using 2.35 now.

**Comment by giordano** [comment]
Can you please update https://github.com/JuliaPackaging/Yggdrasil/blob/master/RootFS.md?  In particular the list of architectures in the first section, and the version of glibc for riscv64 in corresponding table.

File: 0_RootFS/gcc_common.jl [hunk 2]
```diff
@@ -352,6 +357,12 @@ function gcc_script(compiler_target::Platform)
             GLIBC_CONFIGURE_OVERRIDES+=( libc_cv_ssp=no libc_cv_ssp_strong=no )
         fi
 
+        if [[ ${COMPILER_TARGET} == riscv64-* ]]; then
+            # Explicitly disable C++
+            # (Disable for all architectures?)
+            GLIBC_CONFIGURE_OVERRIDES+=( CXX=false )
+        fi
```
Comment (giordano) [inline 0_RootFS/gcc_common.jl:364]: Why this?  What are the consequences?

Comment (eschnett) [reply 0_RootFS/gcc_common.jl:364]: Without this setting, glibc builds the file `support/links-dso-program.cc` with a C++ compiler. There is no C++ compiler for the host architecture available, so glibc uses `/usr/bin/g++` instead. This fails. If we disable C++ (`/bin/false` always reports an error in the configure stage), then glibc builds a different version of that file (`support/links-dso-program-c.c`) with a C compiler, and all is fine.

Comment (eschnett) [reply 0_RootFS/gcc_common.jl:364]: This file doesn't exist in the earlier glibc versions that we build for other architectures.

Comment (giordano) [reply 0_RootFS/gcc_common.jl:364]: > There is no C++ compiler for the host architecture available

All this stuff is built with system compilers, we don't use BinayBuilder toolchains, that's the point of these (obscure) lines https://github.com/JuliaPackaging/Yggdrasil/blob/b196211d789606f65ea85a0624c7fd4422c9a51b/0_RootFS/gcc_common.jl#L31-L32  The C file is probably being built with `/usr/bin/gcc`.

**Pull request merged**